_posts/2022-04-26-fivetran-custom-connector.md

```diff
@@ -0,0 +1,335 @@
+
+- Incremental updates
+- Source data type inference
+- Automatic schema updates
+- Data de-deduplication
+- Destination ingestion optimisations
+- Logs and alerts to monitor events and troubleshoot issues
+
+Use FiveTran's Function connectors if:
+
+- Fivetran doesn't have a connector for your source
+- You are using private APIs or custom applications
+- You are using a source or API that Fivetran is unlikely to support in the near future
+- You want to sync unsupported file formats that require pre-processing
+- You have sensitive data that needs filtering or anonymizing before entering the destination
+
+*Fivetran* . Fivetran Custom Connector. (n.d.). Retrieved May 26, 2022, from https://fivetran.com/docs/functions 
+
+<br>
+
+***
+<br>
+## General Principles
+
+**State Management**
+
+State is a JSON object that contains cursors from the previous successful FiveTran executions run.
+
+[https://fivetran.com/docs/functions/faq/use-state-object](https://fivetran.com/docs/functions/faq/use-state-object) 
+
+[https://fivetran.com/docs/functions/faq/use-secrets-object](https://fivetran.com/docs/functions/faq/use-secrets-object)
+
+
+**Guardrails to consider**
+
+- Lambda timeouts and payload limit
+- Ensure calls to the Lambda are **idempotent** for a given FiveTran cursor
+- Be wary or avoid using cursors supplied by the pagination of an upstream API that may not be persistent.
+
+
+**FiveTran Pagination**
+
+FiveTran pagination allows your lambda function to specify if there is more data to be collected, this is achieved using the hasMore boolean return.
+
+When hasMore = True ,state is updated as normal, however FiveTran immediately calls the lambda with the updated state. This will keep occurring until hasMore = False, which then resets FiveTran to its default state.
+
+[https://fivetran.com/docs/functions/faq/use-hasmore-flag](https://fivetran.com/docs/functions/faq/use-hasmore-flag)
+
+Example Expected function payload
+
+```
+{
+      state: {
+        cursor: '2020-01-01',
+        paginationCounter: 0
+      },
+      secrets: Object
+    }
+```
+
+Example Expected Connector response format
+
+```
+state: {
+cursor: cursorPoint,
+paginationCounter: 1
+    },
+    insert: {
+      Table: apiResponseJson
+    },
+    schema: {
+      Table: {
+        primary_key: ['some_unique_key']
+      }
+    },
+    hasMore: boolean
+  }
+}
+
+```
+
+<br/>
+
+***
+
+<br/>
+## Sequence of FiveTran Connector Excecutions
+
+![](/img/five_tran_sequence.png)
+
+
+For the sake of this example and readability, the return JSON from the api call will be summarised as *apiJsonResponse*
+
+### Call 1: Initial Sync
+
+
+**FiveTranCall**
+
+```
+state: {}
+```
+
+**Connector Response**
+
+```
```

Comment:
format the json for readability,

you might also be able [to style it](https://stackoverflow.com/questions/14901245/how-to-style-a-json-block-in-github-wiki) (YMMV)

Reply:
I originally had it as json but didn't like the way it looked, but that link gives more options, will implement

Reply:
updated